### ISSUE TITLE
Guidance on using `mirai` inside of a package

### ISSUE THREAD

**Issue by andrewGhazi** [shikokuchuo/mirai#137]
Hi there, I've been learning about `mirai` and wanted to ask if there are any recommendations on how to use it from within a package (sort of like [this futureverse vignette](https://future.futureverse.org/articles/future-7-for-package-developers.html)). Apologies if this exists and I just missed it, but I was left with some lingering questions after looking over the pkgdown site and the GitHub issues.

Ideally I'd like my main function to provide a `.cores` argument that handles all the parallelization details for my users when set to 2 or more, but maybe that's inconsistent with the design of `mirai`.

* Should I the developer set up the daemons or should I leave that to the user?
* How should I invoke `mirai` from within my package? I'm not sure if replacing all instances of `lapply()` with `mirai_map()` would work. Ideally I wouldn't need to spawn any additional R processes if the user doesn't want parallelization. I'd prefer to find some way to use `lapply()` by default, but use `mirai_map()` when the user sets `.cores ` to 2 or more. I guess I could try a custom map function like this that `lapply()`s if `.cores == 1` and `mirai_map()`s otherwise, but this seems pretty hacky.

```{r}
my_map = function(.x, .f, ..., .cores = 1L) {
  if (.cores == 1) {
    lapply(.x, .f, ...)
  } else {
    mirai_map(.x, .f, ..., 
              .args = list(other_pkg_fn1 = f1, 
                           other_pkg_fn2 = f2))[]
  }
}
```

* The other option here might be just to export a second version of my function that's identical but uses `mirai_map()`. The downside here is code duplication of course. 
* What's the best way to make other functions from my package available on the daemons? Would I have to run `everywhere()` for each one or could I do `everywhere(library(mypackage))`? The latter seems dangerous if my package is handling the daemon setup.
* My package includes some C++ functions that I run with Rcpp. Are there any special considerations I should be aware of on that front?

If it's helpful, my package is [here](https://github.com/andrewGhazi/fsummary).

Apologies if these are naive questions, I'm still internalizing the design of `mirai`. It seems like really excellent work, the performance in my tests so far is amazing.

**Comment by shikokuchuo** [comment]
> Hi there, I've been learning about mirai and wanted to ask if there are any recommendations on how to use it from within a package (sort of like [this futureverse vignette](https://future.futureverse.org/articles/future-7-for-package-developers.html)). Apologies if this exists and I just missed it, but I was left with some lingering questions after looking over the pkgdown site and the GitHub issues.

Such as document does not exist, partly as `mirai` as a package/framework is not so 'opinionated'. But perhaps there could be a FAQ for package developers down the line if some of these answers prove useful.

> Ideally I'd like my main function to provide a .cores argument that handles all the parallelization details for my users when set to 2 or more, but maybe that's inconsistent with the design of mirai.

If the function is designed to be run with the number of cores set each time, and you don't mind the cost of launching new daemons then this seems straightforward. You could even use a `with(daemons(.cores), { })` type of construct. Alternatively call `daemons(.cores)` once upfront to set up persistent daemons for the entire session.

> Should I the developer set up the daemons or should I leave that to the user?

This would be up to you and the design of your package. If the intention is for the user to also be using other functions or packages using `mirai` then letting them call `mirai::daemons()` directly may be a good idea. It would also allow the flexibility to connect to remote machines.

This is most likely not the case here, but just so you're aware, if the design is for your daemons to be completely separate for some special purpose, then you could set the compute profile '.compute' to a unique value for your package. Then within the package make mirai calls using this same '.compute'. This would then operate independently of the user using `mirai`, or another package that uses `mirai`.

> The other option here might be just to export a second version of my function that's identical but uses mirai_map(). The downside here is code duplication of course.

`mirai_map()` is designed to be async, just like `mirai()` i.e. it returns immediately and doesn't block. Hence it doesn't make sense to have a fallback to a synchronous `lapply()` in the way it might for a framework that just does parallelization. Either of the options you mention seem fine.

> What's the best way to make other functions from my package available on the daemons? Would I have to run everywhere() for each one or could I do everywhere(library(mypackage))? The latter seems dangerous if my package is handling the daemon setup.

You could do `everywhere(library(mypackage))` and this wouldn't be 'dangerous' as long as the package doesn't do a lot of stuff in its onload hook. Always making namespaced calls i.e. `mypacakge::myfunc()` is probably a better general solution. You can also look at the 'cleanup' argument of `daemon()`, which can be passed via the `daemons()` call  - you can configure it not to clean up packages between evaluations to save on overhead for example.

This also ties in to your earlier question of whether to set up the daemons or leave that to the user - if the user is potentially sharing the daemons for purposes other than your package, then you'd want to make each evaluation self-contained and keep cleanup on.

> My package includes some C++ functions that I run with Rcpp. Are there any special considerations I should be aware of on that front?

Nothing extra here. But in case those use openMP or some threaded parallelism, make sure they don't _each_ max out your cores!

> If it's helpful, my package is [here](https://github.com/andrewGhazi/fsummary).

Happy to answer any specific questions as you're building out. Good luck!

**Comment by cdriveraus** [comment]
including an argument to everywhere() that takes e.g. a vector of compute profile names would be a useful addition for within packages, I think.

**Comment by andrewGhazi** [comment]
Hi there, just dropping back into this thread to lay out my thoughts after the past couple weeks of using `mirai` inside of a package. It's worked out pretty well (`parallel` actually slowed my function down while `mirai` scales nicely with the number of daemons) but I ran into some speedbumps that I thought I'd share. If you'd like me to break any of these off into a separate issue I'd be happy to.

* The 2D multiple map behavior of `mirai_map()` bit me a few times. `list_of_data_frames |> lapply(df_function)` is a common pattern for me, but the 2D mapping prevents me from simply replacing that `lapply` with `mirai_map`. After playing around with different options I ultimately just looked at the source code of mirai_map and adapted the part that doesn't do the 2D mapping. If you're accepting functionality requests, I would humbly suggest adding a optional argument to `mirai_map()` that lets users set `.multi_map = FALSE` (defaulting to `TRUE` to preserve the current behavior) that gets &&'d with the existing condition `cond`.
* I initially had issues calling internal functions from my package from the daemons. I ultimately settled on using a triple colon namespaced call e.g. `mypkg:::myfun()`. The one downside I see of using the triple colon is that it adds a warning to checks: 

```
❯ checking dependencies in R code ... WARNING
<other warnings...>
  There are ::: calls to the package's namespace in its code. A package
    almost never needs to use ::: for its own objects:
    <list of functions from my package where I use :::>
```

I couldn't find much guidance in 'Writing R Extensions' about when it *is* appropriate to use triple colon for a package's own namespace, so maybe this is one of those rare cases? Anyway, just thought I'd bring this up if there are thoughts on this, or if there are recommendations on better ways to access my package's internal functions on the daemons. I guess the other option would be to `system.file()` to find my package's installation then then `source()` the R files there, but without having tried it that feels pretty hacky.

* I couldn't find a usage pattern that worked for me with a `.cores` argument, so I ultimately settled on something sort of like this for my package's main function:

```
multiple_daemons = function() {
  is.matrix(mirai::status()$daemons) && nrow(mirai::status()$daemons) > 1
}

main_fun = function(input) {
  if (multiple_daemons()) {
    chunks = split(input) # structure chunks so they don't trigger multiple mapping
    mirai_map(chunks, .main_fun)[] |> data.table::rbindlist()
  } else {
    .main_fun(input)
  }
}

.main_fun = function(input) {
  # <do stuff>
  return(result_df)
}

```

and I included a package startup message pointing out `mirai::daemons(4)` to users. That way they can have control over how the daemons are set up. I was thinking this would give a similar usage pattern to how `future::plan()` is used.

* Finally I wanted to ask about submission to CRAN. I'm guessing that if I have an example that tries to start up daemons on CRAN's servers the checks will error out and the submission will rejected. Should I just wrap those in `if(interactive()) {}`? Are there any other mirai-on-CRAN issues I should consider?

**Comment by shikokuchuo** [comment]
Thanks for the feedback!

1)
> The 2D multiple map behavior of mirai_map() bit me a few times. 

Would you open a new discussion (in the 'discussions' section) about the map - I'd like to understand what you're facing a bit more - if you can show me examples of what happens vs. what you'd like to happen.

2)
> I couldn't find much guidance in 'Writing R Extensions' about when it _is_ appropriate to use triple colon for a package's own namespace, so maybe this is one of those rare cases? 

It's not an issue at all if the package is not destined for CRAN. For CRAN, you can employ a workaround such as:
```r
ns <- getNamespace("my_pkg_name")
ns$my_func() # call you package functions like this
```
Note: do this only for your own package. You shouldn't access non-exported functions from other packages, and it would be against the spirit of CRAN rules (as well as good collaborative norms).

3)
> ```
> multiple_daemons = function() {
>   is.matrix(mirai::status()$daemons) && nrow(mirai::status()$daemons) > 1
> }
> 

Please don't rely on the shape of the status matrix. If you want something robust, I can suggest:
```r
!is.null(nextget("n")) && nextget("n") > 1L
```
The 'n' option for `nextget()` seems to be missing from the current docs - I will add this in. Alternatively:
```r
!identical(status()$daemons, 0L)
```
I shouldn't think it matters too much if there is precisely one daemon or more than one.

4)
>   * Finally I wanted to ask about submission to CRAN. I'm guessing that if I have an example that tries to start up daemons on CRAN's servers the checks will error out and the submission will rejected. Should I just wrap those in `if(interactive()) {}`? Are there any other mirai-on-CRAN issues I should consider?

Yes, make sure you don't go over CRAN's limits. For the examples, wrapping in `if (interactive) {}` is a solution, or put in comments how to run things in parallel. Also for tests there's nothing to stop you running a full suite of tests locally or on CI, but I'd suggest if you need to test on CRAN to limit those to `daemons(1L, dispatcher = FALSE)` so there is only one additional process, and to always clean up with `daemons(0L)` along with at least a 1s sleep.

**Comment by andrewGhazi** [comment]
> > The 2D multiple map behavior of mirai_map() bit me a few times.
> 
> Would you open a new discussion (in the 'discussions' section) about the map - I'd like to understand what you're facing a bit more - if you can show me examples of what happens vs. what you'd like to happen.

Sure, I opened one here #147 .


> ```r
> ns <- getNamespace("my_pkg_name")
> ns$my_func() # call you package functions like this
> ```
> 
> Note: do this only for your own package. You shouldn't access non-exported functions from other packages, and it would be against the spirit of CRAN rules (as well as good collaborative norms).

Okay thanks, I'll give this a shot. Yes, this would definitely only be for my own package's non-exported functions. 

> Please don't rely on the shape of the status matrix. If you want something robust, I can suggest:
> 
> ```r
> !is.null(nextget("n")) && nextget("n") > 1L
> ```

Okay, thanks. This is  exactly the kind of thing I wanted to check.

> I shouldn't think it matters too much if there is precisely one daemon or more than one.

In my (non-thorough, possibly unfair) tests it was a bit slower to perform a task on one daemon than it was to simply run it on the host process, so in my case I only use mirai if there's more than one daemon available.
 
> Yes, make sure you don't go over CRAN's limits. For the examples, wrapping in `if (interactive) {}` is a solution, or put in comments how to run things in parallel. Also for tests there's nothing to stop you running a full suite of tests locally or on CI, but I'd suggest if you need to test on CRAN to limit those to `daemons(1L, dispatcher = FALSE)` so there is only one additional process, and to always clean up with `daemons(0L)` along with at least a 1s sleep.

Great, this is what I wanted to know.

**Comment by shikokuchuo** [comment]
> > I shouldn't think it matters too much if there is precisely one daemon or more than one.
> 
> In my (non-thorough, possibly unfair) tests it was a bit slower to perform a task on one daemon than it was to simply run it on the host process, so in my case I only use mirai if there's more than one daemon available.


What I meant was rather if the user specifically chooses to set up 1 daemon, then I would just accept their wish to run things on the daemon (regardless of speed). Perhaps they want to test things out before doing a large run over a cluster.

**Comment by shikokuchuo** [comment]
@andrewGhazi I'm about to merge an additional section to the mirai vignette which adds the key points from our conversation above. Feel free to open new issues or discussions if you come across anything else.

Note - there is one additional item I've added. I think the approach of leaving it to users to call `daemons()` is the correct one in the vast majority of cases, but I've also added a new argument 'force' at `daemons()`, which should be set to FALSE by package authors if using within their packages. This point is mentioned in the vignette.
